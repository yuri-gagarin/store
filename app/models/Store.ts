import mongoose, { Schema, Document } from "mongoose";
import { IStorePicture } from "./StorePicture";
type IStorePictureRef = mongoose.Types.ObjectId | IStorePicture;

export interface IStore extends Document {
  title: string;
  description: string;
  images: IStorePictureRef[];
  createdAt: Date;
  editedAt?: Date;
}

const StoreSchema: Schema = new Schema({
  title: {
    type: String,
    required: true
  },
  description: {
    type: String,
    required: true
  },
  images: [
    {
      type: Schema.Types.ObjectId,
      ref: "StorePicture"
    }
  ],
  createdAt: {
    type: Date,
    default: Date.now,
    required: true
  },
  editedAt: {
    type: Date
  }
});
const Store = mongoose.model<IStore>('Store', StoreSchema);
export default Store;