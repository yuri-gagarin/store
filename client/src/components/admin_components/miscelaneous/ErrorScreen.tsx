import React from 'react'
import { Button, Dimmer, Header, Icon, Segment } from 'semantic-ui-react'
import axios from "axios";

type RetryLastReqFunc = <T> () => T
interface Props {
  errorMessage?: string;
  lastRequest?: () => Promise<void>;
}
const ErrorScreen: React.FC<Props> = ({ errorMessage, lastRequest }): JSX.Element => {

  const handleRetry = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (lastRequest) {
      lastRequest();
    }
  };
  
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active>
          <Header as='h2' icon inverted>
            <Icon name='exclamation circle' />
              An Error Occured!
            <Header.Subheader>Dimmer sub-header</Header.Subheader>
            <Button 
              id="errorScreenRetryButton"
              inverted 
              color="green" 
              content="Retry" 
              onClick={handleRetry} 
            />
          </Header>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default ErrorScreen