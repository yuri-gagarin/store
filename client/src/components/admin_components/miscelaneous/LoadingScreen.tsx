import React from 'react'
import { Dimmer, Loader, Segment } from 'semantic-ui-react'

const LoaderExampleText: React.FC<{}> = (props): JSX.Element => {
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active inverted>
          <Loader>Please Wait</Loader>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default LoaderExampleText