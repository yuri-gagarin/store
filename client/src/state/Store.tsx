import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

