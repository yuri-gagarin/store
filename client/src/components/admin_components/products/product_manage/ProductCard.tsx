import React, { useState } from "react";
import { Button, Confirm, Grid } from "semantic-ui-react";
// additional components //
import EditControls from "./EditControls"
// css imports //
import "./css/productCard.css";
// actions and state /
import { deleteProduct } from "../actions/APIProductActions";
import { setCurrentProduct, clearCurrentProduct } from "../actions/UIProductActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps } from "react-router-dom";
import { AdminProductRoutes } from "../../../../routes/adminRoutes";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";

interface Props extends RouteComponentProps {
  product: IProductData;
  imageCount: number;
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const ProductCard: React.FC<Props> = ({ product, imageCount, state, dispatch, history }): JSX.Element => {
  const [ editing, setEditing ] = useState<boolean>(false);
  const [ confirmDeleteOpen, setConfirmDeleteOpen ] = useState<boolean>(false);

  const baseUrl = AdminProductRoutes.MANAGE_ROUTE;
  const { _id, name, price, description, images, createdAt, editedAt } = product;

  const handleProductOpen = (e: React.MouseEvent<HTMLButtonElement>): void => {
    history.push(baseUrl + "/view");
  }
  const handleProductEdit = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (!editing) {
      setCurrentProduct(_id, dispatch, state);
      history.push(baseUrl + "/edit");
      setEditing(true);
    } else {
      clearCurrentProduct(dispatch);
      history.push(baseUrl);
      setEditing(false);
    }
  };
  // delete product click and popup //
  const handleDeleteClick = (): void => {
    setConfirmDeleteOpen(true);
  };
  const confirmProductDeleteAction = (): void => {
    const { _id: productId } = product;
    deleteProduct(productId, dispatch, state)
      .then((_) => {
        // action successful //
      })
      .catch((_) => {
        // handle an error //
      });
  };
  const cancelStoreItemDeleteAction = (): void => {
    setConfirmDeleteOpen(false);
  };
 
  return (
    <React.Fragment>
      <Grid.Row style={{ padding: "0.5em", marginTop: "1em" }}>
        <Confirm 
          open={confirmDeleteOpen}
          onCancel={cancelStoreItemDeleteAction}
          onConfirm={confirmProductDeleteAction}
        />
        <div className="productManageDesc">
          <h3>Name: {name}</h3>
          <p><strong>Description:</strong> {description}</p>
          <span>Number of Images: <strong>{imageCount}</strong></span>
          <p><span>Created At: </span>{ConvertDate.international(createdAt)}</p>
        </div> 
        <div className="productManageCtrls">
          <Button 
            inverted
            color="green"
            className="productCardBtn" 
            content="Open" 
            onClick={handleProductOpen} 
          />
          <Button 
            inverted
            color="orange"
            className="productCardEditBtn" 
            content="Edit" 
            onClick={handleProductEdit}  
          />
          <Button 
            inverted
            color="red"
            className="productCardDeleteBtn" 
            content="Delete" 
            onClick={handleDeleteClick}
          />
        </div> 
      </Grid.Row>
      {  editing ?  <EditControls handleProductEdit={handleProductEdit}/>: null }
    </React.Fragment>
    
  );
};

export default withRouter(ProductCard);