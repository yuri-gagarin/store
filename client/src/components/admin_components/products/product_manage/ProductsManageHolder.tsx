import React, { useContext, useEffect, useRef, useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import ProductFormHolder from "../forms/ProductFormContainer";
import ProductCard from "./ProductCard";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
import ErrorScreen from "../../miscelaneous/ErrorScreen";
// actions and state //
import { getAllProducts } from "../actions/APIProductActions";
import { Store } from "../../../../state/Store";
// client routing //
import { withRouter, RouteComponentProps, useRouteMatch, Route } from "react-router-dom";
import { AdminProductRoutes } from "../../../../routes/adminRoutes";

interface Props extends RouteComponentProps {
};

const ProductsManageHolder: React.FC<Props> = ({ history }): JSX.Element => {
  const { state, dispatch } = useContext(Store);
  const { loading, loadedProducts, error } = state.productState;
  // local state //
  const [ newDataLoaded, setNewDataLoaded ] = useState<boolean>(false);
  const newProductsRef = useRef(loadedProducts);
  // routing //
  const match = useRouteMatch(AdminProductRoutes.MANAGE_ROUTE)
  const handleBack = () => {
    history.goBack();
  };

  // lifecycle hooks //
  useEffect(() => {
    let isMounted = true;
    if (isMounted) {
      getAllProducts(dispatch)
        .then((_) => {
          setNewDataLoaded(true);
        })
        .catch((_) => {
          setNewDataLoaded(false);
        });
    }
    return () => { isMounted = false };
  }, [ dispatch ]); 

  useEffect(() => {
    if (newProductsRef.current != loadedProducts && !loading && !error) {
      setNewDataLoaded(true);
    } 
  }, [ newProductsRef.current, loadedProducts, loading, error ]);
 
  return (
    newDataLoaded ?
    <Grid id="productsManageHolder" padded stackable columns={2}>
      <Route path={match?.url + "/edit"}> 
        <Grid.Row>
          <Grid.Column computer={12} tablet={6} mobile={16}>
            <h3>Editing Product: { state.productState.currentProductData.name }</h3>
            <Button 
              id="adminProductsManageBackBtn"
              inverted 
              color="green" 
              content="Back" 
              onClick={handleBack}>

            </Button>
          </Grid.Column>
        </Grid.Row>
        <ProductFormHolder />
      </Route>
      <Route exact path={match?.url}>
        <Grid.Row>
          <Grid.Column computer={12} tablet={8} mobile={16}>
          {
            loadedProducts.map((product) => {
              return (
                <ProductCard 
                  key={product._id}
                  product={product}
                  imageCount={product.images.length}
                  state={state}
                  dispatch={dispatch}
                />
              );
            })
          }
          </Grid.Column>
          <Grid.Column computer={4} tablet={8} mobile={16}>
            
          </Grid.Column>
        </Grid.Row>
      </Route>
    </Grid>
    : 
    ( 
      error ? <ErrorScreen lastRequest={ () => getAllProducts(dispatch) }/> : <LoadingScreen /> 
    )
  );
};

// export for testing //
export { ProductsManageHolder };
// default export //
export default withRouter(ProductsManageHolder);