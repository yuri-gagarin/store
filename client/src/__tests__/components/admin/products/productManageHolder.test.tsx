import React from "react";
import { Grid } from "semantic-ui-react";
// test dependencies //
import moxios from "moxios";
import { mount, ReactWrapper } from "enzyme";
import { act } from "react-dom/test-utils";
import { MemoryRouter as Router } from "react-router-dom";
// components //
import ProductsManageHolder from "../../../../components/admin_components/products/product_manage/ProductsManageHolder";
import LoadingScreen from "../../../../components/admin_components/miscelaneous/LoadingScreen";
import ProductCard from "../../../../components/admin_components/products/product_manage/ProductCard";
import ErrorScreen from "../../../../components/admin_components/miscelaneous/ErrorScreen";
// helpers and state //
import { TestStateProvider } from "../../../../state/Store";

describe("Product Manage Holder Tests", () => {
  let products: IProductData[];
  
  beforeAll(() => {
    products = [
      {
        _id: "1",
        name: "name",
        price: "100",
        details: "details",
        description: "description",
        images: [],
        createdAt: "now"
      },
      {
        _id: "2",
        name: "name",
        price: "200",
        details: "details",
        description: "description",
        images: [],
        createdAt: "now"
      }
    ];
  });

  describe("Default Component state at first render", () => {
    let component: ReactWrapper; let loadingScreen: ReactWrapper;

    beforeAll( async () => {
      moxios.install();
      component = mount(
        <Router keyLength={0} initialEntries={["/admin/home/my_products/manage"]}>
          <TestStateProvider>
            <ProductsManageHolder />
          </TestStateProvider>
        </Router>
      );

      await act( async () => {
        await moxios.stubRequest("/api/products", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            products: []
          }
        });
      });
    });
      afterAll(() => {
      moxios.uninstall();
    }); 
    
    it("Should correctly render", () => {
      const manageHolder = component.find(ProductsManageHolder);
      expect(manageHolder).toMatchSnapshot();
    });
    
    it("Should render a 'LoadingScreen' Component before an API call resolves", () => {
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1)
    });
  });
    
  // mock successful API call render tests //
  describe("State after a successful API call", () => {
    let component: ReactWrapper; let loadingScreen: ReactWrapper;

    beforeAll( async () => {
      moxios.install();

      component = mount(
        <Router keyLength={0} initialEntries={["/admin/home/my_products/manage"]}>
          <TestStateProvider>
            <ProductsManageHolder />
          </TestStateProvider>
        </Router>
      );
      
      await act( async () => {
        await moxios.stubRequest("/api/products", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            products: products
          }
        });
      });
    });

    afterAll(() => {
      moxios.uninstall();
    }); 

    it("Should correctly render the initial Loading Screen", () => {
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should not render the initial Loading Screen after the API call", async () => {
      component.update();
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    });
    it("Should NOT render the ErrorScreen Component", () => {
      const errorScreenComponent = component.find(ErrorScreen);
      expect(errorScreenComponent.length).toEqual(0);
    });

    it("Should render the correct ProductManageHolder Component", () => {
      const productManageHolderComp = component.find(ProductsManageHolder).find(Grid);
      expect(productManageHolderComp.length).toEqual(1);
      expect(component.find(ProductsManageHolder)).toMatchSnapshot();
    });
  
    it("Should render correct number of ProductCard components", () => {
      const productCards = component.find(ProductCard);
      expect(productCards.length).toEqual(products.length);
    })
  });
  // END mock successfull API call render tests //
  // mock ERROR API call render tests //
  describe("State after a Error in API call", () => {
    let component: ReactWrapper; let loadingScreen: ReactWrapper;
    const error = new Error("Error occured");

    beforeAll(async () => {
      await act(async () => {
        moxios.install();
        component = await mount(
          <Router keyLength={0} initialEntries={["/admin/home/my_products/manage"]}>
            <TestStateProvider>
              <ProductsManageHolder />
            </TestStateProvider>
          </Router>
        );
        moxios.stubRequest("/api/products", {
          status: 500,
          response: {
            responseMsg: "Error here",
            error: error
          }
        });
      });
      moxios.uninstall();
    });
    afterAll(() => {
      moxios.uninstall();
    });

    it("Should correctly render the initial Loading Screen", () => {
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });
    it("Should not render the initial Loading Screen after an  API call", () => {
      component.update();
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    });
    it("Should render the ErrorScreen Component", () => {
      const errorScreenComponent = component.find(ErrorScreen);
      expect(errorScreenComponent.length).toEqual(1);
    });
    it("Should NOT render the ProductManageHolder Component", () => {
      const productManageHolderComp = component.find(ProductsManageHolder).find(Grid);
      expect(productManageHolderComp.length).toEqual(0);
    });
    it("Should NOT render ANY ProductCard components", () => {
      const productCards = component.find(ProductCard);
      expect(productCards.length).toEqual(0);
    });
    it("Should have a retry Product API call Button", () => {
      const retryButton = component.find(ErrorScreen).render().find("#errorScreenRetryButton");
      expect(retryButton.length).toEqual(1);
    });
    it("Should correctly re-dispatch the 'getProducts' API request with the button click", async () => {
  
      await act( async () => {
        moxios.install();
        moxios.stubRequest("/api/products", {
          status: 200,
          response: {
            responseMsg: "All Ok",
            products: products
          }
        });
        const retryButton = component.find("#errorScreenRetryButton");
        retryButton.at(0).simulate("click");
      });
      component.update();
      const errorScreen = component.find(ErrorScreen);
      const productManageHolderComp = component.find(ProductsManageHolder).find(Grid);
      expect(errorScreen.length).toEqual(0);
      expect(productManageHolderComp.length).toEqual(1);
    });
    it("Should render correct number of 'ProductCard' Components", () => {
      const productCards = component.find(ProductCard);
      expect(productCards.length).toEqual(products.length);
    });
  });
  // END mock successfull API call tests //
});