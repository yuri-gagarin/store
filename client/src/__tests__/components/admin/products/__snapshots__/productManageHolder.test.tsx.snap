// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Manage Holder Tests Default Component state at first render Should correctly render 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "q8pk0y",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "q8pk0y",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Component>
      <withRouter(ProductsManageHolder)>
        <ProductsManageHolder
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "q8pk0y",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "q8pk0y",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "key": "q8pk0y",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <LoaderExampleText>
            <div
              style={
                Object {
                  "height": "100vh",
                }
              }
            >
              <Segment
                style={
                  Object {
                    "height": "100vh",
                  }
                }
              >
                <div
                  className="ui segment"
                  style={
                    Object {
                      "height": "100vh",
                    }
                  }
                >
                  <Dimmer
                    active={true}
                    inverted={true}
                  >
                    <DimmerInner
                      active={true}
                      inverted={true}
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": null,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": null,
                            }
                          }
                        >
                          <div
                            className="ui active transition visible inverted dimmer"
                            onClick={[Function]}
                          >
                            <div
                              className="content"
                            >
                              <Loader>
                                <div
                                  className="ui text loader"
                                >
                                  Please Wait
                                </div>
                              </Loader>
                            </div>
                          </div>
                        </RefFindNode>
                      </Ref>
                    </DimmerInner>
                  </Dimmer>
                </div>
              </Segment>
            </div>
          </LoaderExampleText>
        </ProductsManageHolder>
      </withRouter(ProductsManageHolder)>
    </Component>
  </Router>
</MemoryRouter>
`;

exports[`Product Manage Holder Tests State after a successful API call Should render the correct ProductManageHolder Component 1`] = `null`;
